File: agents/livekit/lib/aux-stt.ts
/**
 * Auxiliary ("second opinion") speech recognition — `options.stt.aux`.
 *
 * Runs a second, independent STT engine over the caller's audio alongside the
 * agent's own recognition (the pipeline STT, or a realtime model's built-in
 * transcription) and logs each final transcript it produces as a `user-aux`
 * transaction-log entry next to the primary `user` entry, so two recognitions
 * of the same speech can be compared. The auxiliary engine never feeds the
 * model — it is observation only.
 *
 * Mechanism: the worker already holds its own rtc-node connection to the room,
 * so this opens one extra {@link AudioStream} on the caller participant's audio
 * track — the technique bridge-transcription.ts uses for the bridged humans —
 * and pumps it into a fresh STT stream. The AgentSession and this stream consume
 * the same track independently; the session is untouched. STT vendor selection
 * reuses the pipeline resolution with `options.stt.aux` standing in for
 * `options.stt` (`inference.STT.fromModelString(resolvePipelineStt(...))`, or
 * the direct Deepgram plugin under LIVEKIT_PIPELINE_USE_PROVIDER_KEYS), so the
 * same vendor strings mean the same thing in both places.
 *
 * Metering: milliseconds come from the engine's own usage report
 * (`metrics_collected` → `audioDurationMs`, the same event the primary STT
 * meter reads as `stt_metrics`), which both the LiveKit Inference stream and the
 * Deepgram plugin derive from the audio they actually SENT to the vendor — so
 * an engine that never connects (bad credentials, say) meters nothing, however
 * much audio we pumped at it. Final transcripts are counted in characters. Both
 * are reported through `onUsage` and land as `stt-aux` usage rows attributed to
 * the agent call — its own technology, so the second engine's consumption is
 * never merged with, or gated like, the primary `stt` meter (realtime models
 * bundle their own recognition into the model charge; the auxiliary engine is a
 * real extra cost). The audio we streamed is kept separately for diagnostics.
 *
 * Everything here is best-effort: an auxiliary STT failure must never disturb
 * the call. The stream self-disposes when the caller leaves or the room
 * disconnects; the runtime disposes it on teardown, on a full agent handover
 * (re-arming for the incoming agent's own configuration) and when the agent's
 * media is detached after a bridged transfer.
 */

import { inference, stt } from "@livekit/agents";
import { AudioStream, RoomEvent } from "@livekit/rtc-node";
import type { AudioFrame, RemoteParticipant, Room, Track } from "@livekit/rtc-node";
import logger from "./logger.js";
import type { Agent } from "./api-client.js";
import { resolvePipelineStt } from "./pipeline-inference-options.js";
import {
  buildProviderPipelineStt,
  pipelineUsesProviderApiKeys,
} from "./pipeline-provider-keys.js";
import { fromServiceString, type VendorDetail } from "./usage-vendors.js";
import { waitForAudioTrack, waitForRemoteParticipant } from "./bridge-transcription.js";

/** Ledger technology for auxiliary-STT usage rows (distinct from the primary `stt`). */
export const AUX_STT_TECHNOLOGY = "stt-aux";
/** Transaction-log type for auxiliary transcripts (next to the primary `user`). */
export const AUX_STT_LOG_TYPE = "user-aux";
/** Sample rate the caller track is decoded at for the auxiliary stream. */
const AUX_STT_SAMPLE_RATE = 16000;
/**
 * On dispose, how long to give the engine to report its last partial usage
 * interval after we ask it to flush, before the stream is closed for good.
 */
const USAGE_FLUSH_GRACE_MS = 300;
/** Streamed audio below which "the engine returned nothing" is not worth a warning. */
const SILENT_ENGINE_WARN_MS = 3000;

/** Normalised `options.stt.aux`: the same fields as `options.stt`. */
export interface AuxSttConfig {
  vendor?: string;
  language?: string;
}

/**
 * Normalise `options.stt.aux` to null (off) or `{vendor?, language?}`.
 * Lenient — the server validated the shape at save time. Absent / `false` /
 * `enabled: false` / malformed → off; `true` or `{}` → platform defaults.
 */
export function parseAuxSttOption(options: any): AuxSttConfig | null {
  const raw = options?.stt?.aux;
  if (raw === undefined || raw === null || raw === false) return null;
  if (raw === true) return {};
  if (typeof raw !== "object" || Array.isArray(raw) || raw.enabled === false) {
    return null;
  }
  const vendor =
    typeof raw.vendor === "string" && raw.vendor.trim() ? raw.vendor.trim() : undefined;
  const language =
    typeof raw.language === "string" && raw.language.trim()
      ? raw.language.trim()
      : undefined;
  return { ...(vendor ? { vendor } : {}), ...(language ? { language } : {}) };
}

/**
 * The agent with `options.stt.aux` standing in for `options.stt`, so the
 * pipeline STT resolvers build the auxiliary engine exactly as they would the
 * primary one. Language falls back to the agent's own `stt.language`, then
 * `tts.language` (the platform's declare-once convention); the nested `aux`
 * block itself is dropped.
 */
export function auxSttAgent(agent: Agent, config: AuxSttConfig): Agent {
  const options: any = agent?.options || {};
  const inherited =
    (typeof options.stt?.language === "string" && options.stt.language.trim()) ||
    (typeof options.tts?.language === "string" && options.tts.language.trim()) ||
    undefined;
  const language = config.language ?? inherited;
  const sttBlock: Record<string, string> = {};
  if (config.vendor) sttBlock.vendor = config.vendor;
  if (language) sttBlock.language = language;
  return { ...agent, options: { ...options, stt: sttBlock } } as Agent;
}

/** Canonical billing `{vendor, detail}` for the auxiliary engine. */
export function resolveAuxSttVendor(agent: Agent, config: AuxSttConfig): VendorDetail {
  try {
    return fromServiceString(resolvePipelineStt(auxSttAgent(agent, config)));
  } catch {
    const vendor = config.vendor?.split("/")[0]?.split(":")[0]?.trim().toLowerCase();
    return vendor ? { vendor } : {};
  }
}

/**
 * A fresh STT instance for the auxiliary engine — the pipeline's own
 * resolution applied to `options.stt.aux`: LiveKit Inference by default, or
 * the direct Deepgram plugin under LIVEKIT_PIPELINE_USE_PROVIDER_KEYS.
 */
export function buildAuxStt(agent: Agent, config: AuxSttConfig): stt.STT {
  const effective = auxSttAgent(agent, config);
  if (pipelineUsesProviderApiKeys()) {
    return buildProviderPipelineStt(effective);
  }
  return inference.STT.fromModelString(resolvePipelineStt(effective));
}

/** Live auxiliary transcription: usage so far + teardown. */
export interface AuxSttHandle {
  /**
   * Stop the audio pump, ask the engine to report its last usage interval, then
   * close it. Resolves once that report had its chance to land (bounded by
   * {@link USAGE_FLUSH_GRACE_MS}), so a caller that flushes meters right after
   * can await it. Idempotent.
   */
  dispose(): Promise<void>;
  /**
   * `milliseconds` = audio the engine reported accepting (what is metered);
   * `characters` = final transcript text; `streamedMilliseconds` = audio we
   * pumped at it (diagnostic only — it is not billed).
   */
  readonly usage: { milliseconds: number; characters: number; streamedMilliseconds: number };
}

/** Anything the pump can iterate for caller audio (the rtc-node AudioStream, or a test fake). */
export type AuxAudioSource = (AsyncIterable<AudioFrame> | { getReader(): any }) & {
  cancel?: () => Promise<unknown>;
};

export interface ArmAuxSttParams {
  /** The worker's connected RTC room (ctx.room). */
  room: Room;
  /** Server-side room name, for logging. */
  roomName: string;
  /** Identity of the caller participant whose audio is transcribed. */
  callerIdentity: string;
  /** The agent whose `options.stt.aux` (and language defaults) apply. */
  agent: Agent;
  config: AuxSttConfig;
  /** A final transcript from the auxiliary engine, with the time it arrived. */
  onTranscript: (text: string, at: Date) => void;
  /** A usage delta for the aux meter. */
  onUsage: (unit: "milliseconds" | "characters", quantity: number) => void;
  /** Injection points (tests): engine construction, track resolution, audio source. */
  deps?: {
    buildStt?: (agent: Agent, config: AuxSttConfig) => stt.STT;
    resolveTrack?: (room: Room, identity: string) => Promise<Track>;
    openAudioStream?: (track: Track) => AuxAudioSource;
  };
}

async function defaultResolveTrack(room: Room, identity: string): Promise<Track> {
  const participant = await waitForRemoteParticipant(room, identity);
  return waitForAudioTrack(room, participant);
}

function defaultOpenAudioStream(track: Track): AuxAudioSource {
  return new AudioStream(track, {
    sampleRate: AUX_STT_SAMPLE_RATE,
    numChannels: 1,
  }) as unknown as AuxAudioSource;
}

/** Iterate an audio source whether it is a web ReadableStream or a plain async iterable. */
async function* frames(source: AuxAudioSource): AsyncGenerator<AudioFrame> {
  if (typeof (source as any)[Symbol.asyncIterator] === "function") {
    for await (const frame of source as AsyncIterable<AudioFrame>) yield frame;
    return;
  }
  const reader = (source as { getReader(): any }).getReader();
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) return;
      yield value as AudioFrame;
    }
  } finally {
    try {
      reader.releaseLock();
    } catch {
      /* already released */
    }
  }
}

/**
 * Arm the auxiliary STT on the caller's audio track. Resolves the participant
 * and track (waiting for them if needed), builds the engine, then pumps the
 * decoded audio into its stream while forwarding final transcripts and usage.
 * Never throws — transcription is best-effort.
 */
export function armAuxStt(params: ArmAuxSttParams): AuxSttHandle {
  const { room, roomName, callerIdentity, agent, config, onTranscript, onUsage } = params;
  const buildStt = params.deps?.buildStt ?? buildAuxStt;
  const resolveTrack = params.deps?.resolveTrack ?? defaultResolveTrack;
  const openAudioStream = params.deps?.openAudioStream ?? defaultOpenAudioStream;

  const usage = { milliseconds: 0, characters: 0, streamedMilliseconds: 0 };
  let disposed = false;
  let disposing: Promise<void> | null = null;
  const cleanups: Array<() => void> = [];
  /** Asks the live engine to report its last partial usage interval (set once the stream exists). */
  let flushUsageReport: (() => Promise<void>) | null = null;

  const dispose = (): Promise<void> => {
    if (disposing) return disposing;
    disposing = (async () => {
      disposed = true;
      room.off(RoomEvent.ParticipantDisconnected, onParticipantDisconnected);
      room.off(RoomEvent.Disconnected, onRoomDisconnected);
      if (flushUsageReport) {
        try {
          await flushUsageReport();
        } catch {
          /* engine already gone */
        }
      }
      for (const fn of cleanups.splice(0)) {
        try {
          fn();
        } catch (e) {
          logger.debug({ e, roomName }, "auxStt: cleanup step failed");
        }
      }
      if (usage.streamedMilliseconds > 0 && usage.milliseconds === 0) {
        logger.warn(
          { roomName, callerIdentity, ...usage },
          "auxStt: the engine accepted none of the streamed audio (connection or credentials failure?); nothing metered",
        );
      } else if (usage.streamedMilliseconds >= SILENT_ENGINE_WARN_MS && usage.characters === 0) {
        logger.warn(
          { roomName, callerIdentity, ...usage },
          "auxStt: the engine returned no transcript for the streamed audio",
        );
      }
      logger.info(
        { roomName, callerIdentity, ...usage },
        "auxStt: auxiliary transcription disposed",
      );
    })();
    return disposing;
  };

  const onParticipantDisconnected = (p: RemoteParticipant): void => {
    const identity = p?.identity ?? (p as any)?.info?.identity;
    if (identity === callerIdentity) void dispose();
  };
  const onRoomDisconnected = (): void => {
    void dispose();
  };
  room.on(RoomEvent.ParticipantDisconnected, onParticipantDisconnected);
  room.on(RoomEvent.Disconnected, onRoomDisconnected);

  const run = async (): Promise<void> => {
    try {
      const track = await resolveTrack(room, callerIdentity);
      if (disposed) return;

      const sttInstance = buildStt(agent, config);
      const sttStream = sttInstance.stream();
      const audio = openAudioStream(track);

      // Meter what the ENGINE says it accepted, not what we pumped: both the
      // Inference stream and the Deepgram plugin report audio they actually sent
      // to the vendor (the plugin every 5 s, flushed on demand), and report
      // nothing while a connection is failing — so a 401 from the vendor cannot
      // become a bill for the caller.
      const onMetrics = (m: any): void => {
        const ms = Math.round(Number(m?.audioDurationMs) || 0);
        if (ms <= 0) return;
        usage.milliseconds += ms;
        try {
          onUsage("milliseconds", ms);
        } catch (e) {
          logger.debug({ e, roomName }, "auxStt: usage report failed");
        }
      };
      // The SDK surfaces engine failures as events, not throws (a failing
      // connection retries quietly for minutes): log them, and stop on a
      // non-recoverable one rather than keep pumping into a dead stream.
      const onEngineError = (ev: any): void => {
        const recoverable = ev?.recoverable !== false;
        logger.warn(
          { roomName, callerIdentity, label: ev?.label, recoverable, err: ev?.error?.message ?? String(ev?.error) },
          recoverable
            ? "auxStt: engine error (recoverable)"
            : "auxStt: engine failed (not recoverable); stopping the auxiliary transcription",
        );
        if (!recoverable) void dispose();
      };
      sttInstance.on("metrics_collected", onMetrics);
      sttInstance.on("error", onEngineError);
      flushUsageReport = async () => {
        sttStream.flush(); // FLUSH_SENTINEL → the engine reports its last partial usage interval
        await new Promise((resolve) => setTimeout(resolve, USAGE_FLUSH_GRACE_MS));
      };
      const stop = () => {
        try {
          sttInstance.off("metrics_collected", onMetrics);
          sttInstance.off("error", onEngineError);
        } catch {
          /* not an emitter (tests) */
        }
        try {
          sttStream.endInput();
        } catch {
          /* already ended */
        }
        try {
          sttStream.close();
        } catch {
          /* already closed */
        }
        void sttInstance.close().catch(() => {});
        void audio.cancel?.().catch?.(() => {});
      };
      cleanups.push(stop);
      if (disposed) {
        // dispose() ran between the awaits above and the push — unwind now.
        stop();
        return;
      }

      // Pump the caller's audio into the engine ourselves (rather than handing
      // the stream over) so we know how much we streamed — a diagnostic only;
      // the meter is the engine's own report above.
      const pump = (async (): Promise<void> => {
        try {
          for await (const frame of frames(audio)) {
            if (disposed) break;
            const rate = frame.sampleRate || AUX_STT_SAMPLE_RATE;
            usage.streamedMilliseconds += (frame.samplesPerChannel / rate) * 1000;
            try {
              sttStream.pushFrame(frame);
            } catch {
              break; // input ended underneath us (dispose / stream closed)
            }
          }
        } catch (e) {
          if (!disposed) {
            logger.warn({ e, roomName, callerIdentity }, "auxStt: audio pump failed");
          }
        } finally {
          try {
            sttStream.endInput();
          } catch {
            /* already ended */
          }
        }
      })();

      logger.info(
        { roomName, callerIdentity, label: sttInstance.label, config },
        "auxStt: auxiliary transcription armed on the caller track",
      );

      for await (const ev of sttStream) {
        if (disposed) break;
        if (ev.type !== stt.SpeechEventType.FINAL_TRANSCRIPT) continue;
        const text = (ev.alternatives?.[0]?.text ?? "").trim();
        if (!text) continue;
        usage.characters += text.length;
        try {
          onUsage("characters", text.length);
        } catch (e) {
          logger.debug({ e, roomName }, "auxStt: usage report failed");
        }
        try {
          onTranscript(text, new Date());
        } catch (e) {
          logger.warn({ e, roomName, callerIdentity }, "auxStt: transcript handler failed");
        }
      }
      await pump.catch(() => {});
    } catch (e) {
      // Best-effort: the call carries on without the second opinion.
      if (!disposed) {
        logger.warn(
          { e, roomName, callerIdentity },
          "auxStt: auxiliary transcription failed (continuing without it)",
        );
      }
    }
  };
  void run();

  return {
    dispose,
    get usage() {
      return usage;
    },
  };
}

File: agents/pipecat/pipecat_aplisay/aux_stt.py
"""Side STT taps: the auxiliary ("second opinion") recognition of the CALLER
(``options.stt.aux``) and the output audit recognition of the AGENT
(``options.tts.output``).

Runs a second, independent STT engine over the caller's audio alongside the
agent's own recognition (the pipeline STT, or a realtime model's built-in
transcription) and logs each final transcript it produces as a ``user-aux``
transaction-log entry next to the primary ``user`` entry, so two recognitions
of the same speech can be compared. The auxiliary engine never feeds the model
— it is observation only.

Mechanism: an :class:`AuxSttTap` sits right after ``transport.input()`` (behind
the WebRTC relay tap, so an engaged relay silences it too). It passes every
frame through untouched and copies each ``InputAudioRawFrame`` into a side
STT-only pipeline (:class:`~pipecat_aplisay.bridge_transcript.SttStream` — the
same helper the bridged-segment transcription uses), so the auxiliary service
never sits in the main chain: none of its transcription, metrics or settings
frames can reach the context aggregator or the primary STT, and an STT control
frame aimed at the primary engine is never swallowed by the wrong service.

The engine is built by ``voice_session.build_stt_service`` with
``options.stt.aux`` standing in for ``options.stt`` (:func:`aux_stt_agent`), so
the same vendor strings mean the same thing in both places.

Metering: the audio streamed to the auxiliary engine is measured at the tap
(milliseconds, silence included — the basis streaming STT vendors bill on) and
final transcripts are counted in characters. Pipecat's STT services expose no
"audio the vendor accepted" figure, so streamed audio is metered only once the
engine has proved itself by returning a transcript in this call: until then the
milliseconds are held back, and an engine that never produces one (rejected
credentials, a dead connection) meters nothing rather than billing the caller
for a service that delivered nothing. Both meters are reported through
``on_usage`` and land as ``stt-aux`` usage rows — their own technology, so the
second engine's consumption is neither merged with nor gated like the primary
``stt`` meter (a realtime model bundles its own recognition into the model
charge; the auxiliary engine is a real extra cost on every voice mode).

Output audit (``options.tts.output``): the same tap class, pointed at the
agent's own ``TTSAudioRawFrame``s immediately before ``transport.output()`` —
the synthesised (pipeline) or model-produced (realtime) audio the caller
actually hears. Its finals are logged as ``agent-speech`` beside the ``agent``
turn the model produced, and metered as ``stt-output``. Only TTS audio is
tapped: the confidence tone, a fixed fallback message and relayed peer audio are
plain ``OutputAudioRawFrame``s and are ignored.

Neither tap exists on a call unless its option is set: ``voice_session`` builds
a tap only for a configured option, and the engine is built lazily on the first
audio frame inside it.

Everything here is best-effort: a side STT failure must never disturb the call.
See docs/auxiliary-stt.md.
"""

from __future__ import annotations

import asyncio
from typing import Any, Awaitable, Callable, Optional

from loguru import logger
from pipecat.frames.frames import CancelFrame, EndFrame, InputAudioRawFrame, TTSAudioRawFrame
from pipecat.processors.frame_processor import FrameDirection, FrameProcessor

from .bridge_transcript import SttStream

#: Ledger technology for auxiliary-STT usage rows (distinct from the primary ``stt``).
AUX_STT_TECHNOLOGY = "stt-aux"
#: Transaction-log type for auxiliary transcripts (next to the primary ``user``).
AUX_STT_LOG_TYPE = "user-aux"
#: Ledger technology for the output audit's usage rows.
OUTPUT_STT_TECHNOLOGY = "stt-output"
#: Transaction-log type for the output audit's transcripts (next to ``agent``).
OUTPUT_STT_LOG_TYPE = "agent-speech"
#: Default engine when a side option's ``vendor`` is unset — the pipeline default.
DEFAULT_AUX_STT_VENDOR = "deepgram"


def _parse_side_stt_block(raw: Any) -> Optional[dict]:
    """Normalise one side-STT block to ``None`` (off) or
    ``{"vendor": Optional[str], "language": Optional[str]}``. Lenient — the
    server validated the shape at save time: absent / ``False`` /
    ``enabled: False`` / malformed → off; ``True`` or ``{}`` → defaults."""
    if raw is None or raw is False:
        return None
    if raw is True:
        return {"vendor": None, "language": None}
    if not isinstance(raw, dict) or raw.get("enabled") is False:
        return None
    vendor = raw.get("vendor")
    language = raw.get("language")
    return {
        "vendor": vendor.strip() if isinstance(vendor, str) and vendor.strip() else None,
        "language": language.strip() if isinstance(language, str) and language.strip() else None,
    }


def parse_aux_stt_option(options: Optional[dict]) -> Optional[dict]:
    """``options.stt.aux`` → side-STT config or ``None`` (off)."""
    stt_opts = (options or {}).get("stt")
    if not isinstance(stt_opts, dict):
        return None
    return _parse_side_stt_block(stt_opts.get("aux"))


def parse_output_stt_option(options: Optional[dict]) -> Optional[dict]:
    """``options.tts.output`` → side-STT config or ``None`` (off)."""
    tts_opts = (options or {}).get("tts")
    if not isinstance(tts_opts, dict):
        return None
    return _parse_side_stt_block(tts_opts.get("output"))


def _side_stt_agent(agent: dict, config: dict, language_order: tuple[str, ...]) -> dict:
    """The agent with the side block standing in for ``options.stt``, so
    ``build_stt_service`` constructs the side engine exactly as it would the
    primary one. Language falls back through ``language_order`` (the
    platform's declare-once convention); nested side blocks are dropped."""
    options = dict(agent.get("options") or {})
    language = config.get("language")
    for key in language_order:
        if language:
            break
        block = options.get(key)
        language = block.get("language") if isinstance(block, dict) else None
    block: dict = {}
    if config.get("vendor"):
        block["vendor"] = config["vendor"]
    if language:
        block["language"] = language
    options["stt"] = block
    return {**agent, "options": options}


def aux_stt_agent(agent: dict, config: dict) -> dict:
    """Caller side: language falls back to ``stt.language`` then ``tts.language``."""
    return _side_stt_agent(agent, config, ("stt", "tts"))


def output_stt_agent(agent: dict, config: dict) -> dict:
    """Agent side: the agent speaks in its TTS language, so ``tts.language`` first."""
    return _side_stt_agent(agent, config, ("tts", "stt"))


def _side_stt_vendor(effective_agent: dict) -> dict:
    """Canonical ``{vendor, model}`` for a side engine's usage rows: ``vendor``
    is the bare engine name (what the rate lines match on); ``model`` is
    ``vendor/model`` when the vendor string was scoped (``deepgram/nova-3``),
    else ``None``."""
    stt_opts = (effective_agent.get("options") or {}).get("stt") or {}
    raw = str(stt_opts.get("vendor") or DEFAULT_AUX_STT_VENDOR)
    head = raw.split(":", 1)[0]
    vendor = head.split("/", 1)[0].strip().lower()
    model = head.split("/", 1)[1].strip() if "/" in head else None
    return {"vendor": vendor, "model": f"{vendor}/{model}" if model else None}


def aux_stt_vendor(agent: dict, config: dict) -> dict:
    return _side_stt_vendor(aux_stt_agent(agent, config))


def output_stt_vendor(agent: dict, config: dict) -> dict:
    return _side_stt_vendor(output_stt_agent(agent, config))


OnFinal = Callable[[str], Awaitable[None]]
OnUsage = Callable[[str, int], None]


class AuxSttTap(FrameProcessor):
    """Pass-through processor that copies one class of audio frame into a side
    STT-only pipeline and forwards its final transcripts + usage. Instantiated
    twice over: on the caller's ``InputAudioRawFrame``s (the auxiliary STT) and
    on the agent's ``TTSAudioRawFrame``s (the output audit).

    Args:
        stt_factory: builds the side STT service (called once, lazily, on the
            first audio frame so a build failure only costs the side transcript,
            never the call).
        on_final: awaited with each non-empty final transcript.
        on_usage: called with ``("milliseconds", n)`` as audio is streamed to
            the engine and ``("characters", n)`` per final transcript.
        frame_cls: the audio frame class to tap (default: the caller's input).
        label: log prefix (``auxStt`` / ``outputStt``).
        stream_factory: test seam — builds the side pipeline; defaults to
            :class:`SttStream`.
    """

    def __init__(
        self,
        *,
        stt_factory: Callable[[], Any],
        on_final: OnFinal,
        on_usage: Optional[OnUsage] = None,
        frame_cls: type = InputAudioRawFrame,
        label: str = "auxStt",
        stream_factory: Optional[Callable[..., Any]] = None,
        name: Optional[str] = None,
    ) -> None:
        super().__init__(name=name or f"{label}Tap")
        self._frame_cls = frame_cls
        self._label = label
        self._stt_factory = stt_factory
        self._on_final = on_final
        self._on_usage: OnUsage = on_usage or (lambda _unit, _qty: None)
        self._stream_factory = stream_factory or SttStream
        self._stream: Optional[Any] = None
        self._sample_rate: Optional[int] = None
        self._num_channels: Optional[int] = None
        self._failed = False
        self._rate_warned = False
        self._milliseconds: float = 0.0
        self._reported_ms = 0
        self._characters = 0
        # Streamed milliseconds held back until the engine returns its first
        # transcript (see module docstring); dropped if it never does.
        self._pending_ms = 0
        self._proven = False
        self._stop_task: Optional[asyncio.Task] = None

    @property
    def usage(self) -> dict:
        """Audio milliseconds streamed to the engine and characters it returned."""
        return {"milliseconds": int(self._milliseconds), "characters": self._characters}

    async def process_frame(self, frame, direction: FrameDirection):  # noqa: ANN001
        await super().process_frame(frame, direction)
        if isinstance(frame, self._frame_cls) and direction == FrameDirection.DOWNSTREAM:
            await self._tap_audio(frame)
        elif isinstance(frame, (EndFrame, CancelFrame)):
            self._schedule_stop()
        await self.push_frame(frame, direction)

    async def _tap_audio(self, frame: InputAudioRawFrame) -> None:
        if self._failed:
            return
        try:
            if self._stream is None:
                self._sample_rate = frame.sample_rate
                self._num_channels = frame.num_channels
                self._stream = self._stream_factory(
                    self._stt_factory(),
                    self._collect_final,
                    sample_rate=frame.sample_rate,
                    num_channels=frame.num_channels,
                )
                await self._stream.start()
                logger.info(
                    f"{self._label}: side transcription armed "
                    f"({frame.sample_rate} Hz, {frame.num_channels} ch)"
                )
            if frame.sample_rate != self._sample_rate or frame.num_channels != self._num_channels:
                # The side pipeline was started at the first frame's format;
                # a mid-call change would need a rebuild — drop rather than
                # feed the engine audio it will mis-decode.
                if not self._rate_warned:
                    self._rate_warned = True
                    logger.warning(
                        f"{self._label}: audio format changed mid-call; dropping mismatched frames"
                    )
                return
            await self._stream.feed(frame.audio)
            self._milliseconds += frame.num_frames / frame.sample_rate * 1000.0
            self._report_audio()
        except Exception as e:  # noqa: BLE001
            # Best-effort: the call carries on without the second opinion.
            self._failed = True
            logger.warning(f"{self._label}: side transcription failed (continuing without it): {e}")
            self._schedule_stop()

    def _report_audio(self) -> None:
        """Account for streamed audio not yet accounted for (whole milliseconds, no
        drift): metered straight away once the engine has proved itself, held
        back until then."""
        delta = int(self._milliseconds) - self._reported_ms
        if delta <= 0:
            return
        self._reported_ms += delta
        if not self._proven:
            self._pending_ms += delta
            return
        self._emit_usage("milliseconds", delta)

    def _emit_usage(self, unit: str, quantity: int) -> None:
        try:
            self._on_usage(unit, quantity)
        except Exception as e:  # noqa: BLE001
            logger.debug(f"{self._label}: usage report failed: {e}")

    async def _collect_final(self, text: str) -> None:
        text = (text or "").strip()
        if not text:
            return
        if not self._proven:
            # The engine works: meter what was streamed before this first transcript too.
            self._proven = True
            pending, self._pending_ms = self._pending_ms, 0
            if pending:
                self._emit_usage("milliseconds", pending)
        self._characters += len(text)
        self._emit_usage("characters", len(text))
        try:
            await self._on_final(text)
        except Exception as e:  # noqa: BLE001
            logger.warning(f"{self._label}: transcript handler failed: {e}")

    def _schedule_stop(self) -> None:
        """Tear the side pipeline down without holding up the main chain's own
        End/Cancel processing (its stop waits on the runner, bounded)."""
        # P10: mark the tap finished as well as clearing the stream.
        # ``_tap_audio`` rebuilds the side pipeline whenever ``_stream``
        # is None, so a stray audio frame arriving after End/Cancel would
        # start a fresh STT stream that nothing then stops. Frame
        # ordering normally prevents that; this makes it impossible.
        self._failed = True
        stream, self._stream = self._stream, None
        if stream is None:
            return
        self._report_audio()
        self._stop_task = asyncio.create_task(self._stop_stream(stream))

    async def _stop_stream(self, stream: Any) -> None:
        try:
            await stream.stop()
        except Exception as e:  # noqa: BLE001
            logger.debug(f"{self._label}: side pipeline stop raised: {e}")
        if not self._proven and self._pending_ms:
            logger.warning(
                f"{self._label}: the engine returned no transcript for {self._pending_ms} ms of "
                "streamed audio (connection or credentials failure?); nothing metered"
            )
        logger.info(
            f"{self._label}: side transcription stopped "
            f"({int(self._milliseconds)} ms streamed, {self._characters} chars)"
        )
